### DISCUSSION TITLE
Does object destructuring affect Valtio's useSnapshot re-render optimisation

### DISCUSSION THREAD

**Discussion by codebycarlos** [pmndrs/valtio#707] [Q&A]
According to the [documentation](https://valtio.pmnd.rs/docs/api/basic/useSnapshot):

"useSnapshot wraps the Valtio snapshot in an access-tracking proxy, so your component is render optimized, i.e. it will only re-render if keys that it (or its child components) specifically accessed have changed, and not on every single change to the proxy."

And the example..

```
function Counter() {
  const snap = useSnapshot(state)
  return (
    <div>
      {snap.count}
      <button
        onClick={() => {
          // also read from the state proxy in callbacks
          if (state.count < 10) {
            ++state.count
          }
        }}>
        +1
      </button>
    </div>
  )
}
```

My question is, if we replace the above with object destructuring (i.e. ```const { count } = useSnapshot(state)```) would the optimisation still work as intended?

**Comment by codebycarlos** [comment]
Based on [this](https://github.com/pmndrs/valtio/blob/3ad457d45f3f2140b552d2d53fc79f925cae50c1/src/react.ts#L46-L117), it appears the answer is no (doesn't affect optimisation).

**Comment by dai-shi** [comment]
> My question is, if we replace the above with object destructuring (i.e. `const { count } = useSnapshot(state)`) would the optimisation still work as intended?

Yes, it still works as intended (= render optimized.)
